lightning/src/ln/onion_utils.rs

```diff
@@ -935,6 +935,160 @@ pub(crate) fn decode_next_payment_hop<NS: Deref>(
+		total_msat: u64,
+		/// outgoing cltv
+		outgoing_cltv_value: u32,
+		/// Payment secret 
+		payment_secret: [u8; 32],
+		/// The maximum total CLTV that is acceptable when relaying a payment over this hop
+		max_cltv_expiry: u32,
+		/// The minimum value, in msat, that may be accepted by the node corresponding to this hop
+		htlc_minimum_msat: u64,
+		/// Blinding point from intro node
+		intro_node_blinding_point: PublicKey,
+	}
+}
+
+impl std::convert::TryFrom<msgs::InboundOnionPayload> for ReceivedPayment {
+    type Error = ();
+    fn try_from(pld: msgs::InboundOnionPayload) -> Result<Self, Self::Error> {
+		match pld {
+			msgs::InboundOnionPayload::Forward { short_channel_id: _, amt_to_forward: _, outgoing_cltv_value: _ } => {
+				Err(())
+			},
+			msgs::InboundOnionPayload::Receive { payment_data, payment_metadata, keysend_preimage, custom_tlvs, amt_msat, outgoing_cltv_value } => {
+				let (payment_secret, total_msat) = match payment_data {
+					Some(p) => (Some(p.payment_secret.0), Some(p.total_msat)),
+					None => (None, None),
+				};
+				let keysend_preimage = keysend_preimage.map(|p| p.0);
+				Ok(Self::Regular { payment_secret, total_msat, payment_metadata, keysend_preimage, custom_tlvs, amt_msat, outgoing_cltv_value })
+			},
+			msgs::InboundOnionPayload::BlindedReceive { amt_msat, total_msat, outgoing_cltv_value, payment_secret, payment_constraints, intro_node_blinding_point } => {
+				let payment_secret = payment_secret.0;
+				let max_cltv_expiry = payment_constraints.max_cltv_expiry;
+				let htlc_minimum_msat = payment_constraints.htlc_minimum_msat;
+				Ok(Self::Blinded { amt_msat, total_msat, outgoing_cltv_value, payment_secret, max_cltv_expiry, htlc_minimum_msat, intro_node_blinding_point })
+			}
+		}
+    }
+}
+
+/// Received and decrypted onion payment, either of type Receive (for us), or Forward.
+#[derive(Debug)]
+pub enum PeeledPayment {
+	/// This onion payload was for us, not for forwarding to a next-hop.
+	Receive(ReceivedPayment),
+	/// This onion payload to be forwarded to next peer.
+	Forward(ForwardedPayment),
+}
+
+/// Unwrap one layer of an incoming HTLC, returning either or a received payment, or a another
+/// onion to forward.
+pub fn peel_payment_onion<NS: Deref>(
+	onion: &msgs::OnionPacket, payment_hash: PaymentHash, node_signer: &NS,
+	secp_ctx: &Secp256k1<secp256k1::All>
+) -> Result<PeeledPayment, ()> 
+where 
+	NS::Target: NodeSigner,
+{
+	if onion.public_key.is_err() {
+		return Err(());
+	}
+	let shared_secret = node_signer
+		.ecdh(Recipient::Node, &onion.public_key.unwrap(), None)
+		.unwrap()
+		.secret_bytes();
+
+	let hop = decode_next_payment_hop(shared_secret, &onion.hop_data[..], onion.hmac, payment_hash, node_signer).map_err(|_| ())?;
+	let peeled = match hop {
+		Hop::Forward { next_hop_data, next_hop_hmac, new_packet_bytes } => {
+			if let msgs::InboundOnionPayload::Forward{short_channel_id, amt_to_forward, outgoing_cltv_value} = next_hop_data {
```

Comment:
Here and elsewhere, wrap at 100 chars:
```suggestion
			if let msgs::InboundOnionPayload::Forward {
				short_channel_id, amt_to_forward, outgoing_cltv_value
			} = next_hop_data {
```